File: server-java/src/main/java/io/maxi/examples/server/model/Transfer.java
package io.maxi.examples.server.model;

import io.maxi.annotation.MaxiField;
import io.maxi.annotation.MaxiType;

/**
 * Transfer DTO — mirrors the X:Transfer type in shared/sports.mxs.
 *
 * <p>{@code player}, {@code fromTeam}, and {@code toTeam} hold raw object-maps
 * (as returned by {@code DataLoader.getTransfersForPlayer()}) rather than
 * typed {@link Player}/{@link Team} instances. The {@code typeExpr} annotations
 * declare the MAXI reference types so the dumper can collect and emit them as
 * separate P/T records when {@code collectReferences} is enabled.
 */
@MaxiType(alias = "X", name = "Transfer")
public class Transfer {

    @MaxiField(typeExpr = "int")
    public int id;

    @MaxiField(typeExpr = "P")
    public Object player;

    @MaxiField(typeExpr = "T")
    public Object fromTeam;

    @MaxiField(typeExpr = "T")
    public Object toTeam;

    @MaxiField(annotation = "date", required = true)
    public String date;

    @MaxiField(typeExpr = "decimal")
    public Double fee;

    public Transfer() {}
}

File: server-php/src/Model/Transfer.php
<?php

declare(strict_types=1);

namespace App\Model;

use Maxi\Attribute\MaxiField;
use Maxi\Attribute\MaxiType;

/**
 * Transfer DTO — mirrors the X:Transfer type in shared/sports.mxs.
 *
 * player / fromTeam / toTeam hold plain associative arrays (as returned by
 * DataLoader::loadTransfersWithRefs()) rather than nested DTO instances.
 * The typeExpr annotations tell MAXI these are P / T reference types so
 * the dumper can collect and emit them as separate records.
 */
#[MaxiType(alias: 'X', name: 'Transfer')]
class Transfer
{
    public function __construct(
        #[MaxiField(typeExpr: 'int')]
        public int $id = 0,

        /** @var array<string,mixed> Full Player record (from DataLoader join) */
        #[MaxiField(typeExpr: 'P')]
        public array $player = [],

        /** @var array<string,mixed> Full Team record for the source club */
        #[MaxiField(typeExpr: 'T')]
        public array $fromTeam = [],

        /** @var array<string,mixed> Full Team record for the destination club */
        #[MaxiField(typeExpr: 'T')]
        public array $toTeam = [],

        #[MaxiField(annotation: 'date', required: true)]
        public string $date = '',

        #[MaxiField(typeExpr: 'decimal')]
        public ?float $fee = null,
    ) {}
}
